Add initial Gem Repositories doc (#1921)
o3de/o3de.org: content/docs/user-guide/gems/repositories/overview.md

---
linkTitle: Overview
title: Gem Repository Overview
description: Gems Repositories provide a way to share additional Gems.
weight: 100
toc: true
---

Gem Repositories provide a way to discover, acquire and share Gems outside those that come with the default Open 3D Engine.  While most Gem Repositories are hosted online, they can also exist locally on a hard drive which is useful for testing.

## Adding an Existing Gem Repository  

### From the Project Manager Gem Repositories Page

Gem Repositories can be added from the Gem Repositories page in the Project Manager.

> 1. Press the `Add Repository` button. A dialog will open asking for a `Repository URL` or `Local Path`. 

Author: You can access the page from the `Engine` tab or from any of the `Gem Catalog` pages.

    "summary": "Example Repository",
    "additional_info": "Example repository hosted on local http server.",
    "last_updated": "10/26/2021 04:56pm",
    "gems": [
        "http://localhost:8080/ExampleGem/"
    ]
}
```

In the future, the format of the `repo.json` file may be expanded to support projects, templates, and additional Gem repositories.

### Format of gem.json files 

The `gem.json` file for your Gems will need to be modified to include the `origin_uri`, `repo_uri`, `last_updated`, and `sha256` fields.
```
{
    "gem_name": "TestGem",
    "display_name": "Downloadable Test Gem",
    "license": "Apache-2.0 Or MIT",
    "origin": "Test Creator",
    "origin_uri": "http://localhost:8080/TestGem/gem.zip",
    "repo_uri": "http://localhost:8080/", 
    "last_updated": "2021-11-18",
    "sha256":"",
    "type": "Tool",
    "summary": "The Downloadable Test Gem provides an example of a downloadable test Gem.",
    "canonical_tags": [
        "Gem"
    ],
    "user_tags": [
        "Gameplay",
        "Simulation",
        "Utility"
    ],
    "icon_path": "preview.png",
    "requirements": "",
    "documentation_url": "https://o3de.org/",
    "dependencies": []
}
```

> | Field | Description |

Author: Great suggestion! I've updated the gem.json documentation and added a link to it from this page

```
| Field | Description |
| --- | --- |

> | origin_uri | The URI for the `.zip` file containing the Gem archive - this is a direct link to download the Gem `.zip` archive. |

Reviewer: ```suggestion
| origin_uri | The URI for the `.zip` file containing the Gem archive.  This is a direct link to download the Gem's `.zip` archive. |
```

> | repo_uri | The URI for the Gem Repository |

Reviewer: ```suggestion
| repo_uri | The URI for the Gem repository. |
```

> | last_updated | The date this file or Gem was last updated in YYYY-MM-DD, YYYY-MM-DD HH:MM:SS or YYYY-MM-DDTHH:MM:SS format |

Reviewer: ```suggestion
| last_updated | The date this file or Gem was last updated in `YYYY-MM-DD`, `YYYY-MM-DD HH:MM:SS`, or `YYYY-MM-DDTHH:MM:SS` format. |
```

> | sha256 | The SHA256 digest of the `.zip` archive pointed to by the `origin_uri` field - for testing you can omit this field, but it is highly recommended you include it. |

Reviewer: ```suggestion
| sha256 | The SHA-256 digest of the `.zip` archive that the `origin_uri` field points to.  You can omit this field for testing, but we highly recommend including it. |
```

> ## Creating a Gem Repository  

Reviewer: ```suggestion
## Creating a Gem repository using a template 
```

> 
> ### Using a Template

Reviewer: ```suggestion
```
We could collapse these headings into one.

> 1. Create a new gem repo using `o3de.bat`.  This example uses a localhost URL as the repository URI for testing.

Reviewer: ```suggestion
1. Create a new Gem repository using `o3de.bat`.  The following example uses a localhost URL as the repository URI for testing.
```

    ```

>     scripts\o3de.bat create-from-template -tp Templates\GemRepo -dp <repo path on disk> -r ${RepoURI} http://localhost:8080/

Reviewer: ```suggestion
    scripts\o3de.bat create-from-template -tp Templates\GemRepo -dp <repository path on disk> -r ${RepoURI} http://localhost:8080/
```

    ```

> 1. Open the `repo.json` file in a text editor and update all the fields except the gem's list - we'll update that when adding a gem to the repository in the next step.

Reviewer: ```suggestion
1. Open the `repo.json` file in a text editor and update all the fields except the Gems list.  In the next step, you will update that when adding a Gem to the repository.
```

    ```
    {
        "repo_name":"RepositoryName",
        "origin":"RemoteOriginName",
        "repo_uri": "http://localhost:8080/",
        "summary": "Example Repository",
        "additional_info": "Example repository hosted on local http server.",
        "last_updated": "10/26/2021 04:56pm",
        "gems": [
        ]
    }
    ```

> ### Add a Gem to the Gem Repository

Reviewer: ```suggestion
### Add a Gem to the Gem repository
```

> 1.  Make a folder inside your Gem Repository for the Gem.  In this example we'll assume the gem is named "RemoteExample" and make a folder named "RemoteExampleV1" and place a `gem.json` in there that looks like this:

Reviewer: ```suggestion
1.  Make a folder inside your Gem repository for the Gem.  In this example, the gem is named `RemoteExample`. Make a folder named `RemoteExampleV1` and create a `gem.json` file in the folder that looks like this:
```

    ```
    {
        "gem_name": "RemoteExample",
        "display_name": "Example Remote Gem",
        "license": "Apache-2.0 Or MIT",
        "origin":"RemoteOriginName",
        "origin_uri": "http://localhost:8080/RemoteExampleV1/RemoteExampleV1.zip",
        "sha256":"",
        "type": "Code",
        "summary": "An example remote gem.",
        "canonical_tags": [
            "Gem"
        ],
        "user_tags": [
            "Gameplay"
        ],
        "icon_path": "preview.png",
        "requirements": "",
        "documentation_url": "https://o3de.org/docs",
        "dependencies": []
    }
    ```

    The `origin_uri` points to the location of the gem `.zip` archive that will be downloaded

> 1. Create a Gem, or take an existing Gem, zip up the contents and place it in the "RemoteExampleV1" folder, and name it `RemoteExampleV1.zip`

Reviewer: ```suggestion
1. Create a Gem, or take an existing Gem, zip up the Gem's contents, not the Gem folder, and place the `.zip` in the "RemoteExampleV1" folder, and name it `RemoteExampleV1.zip`.
```

> 1. Create a Gem, or take an existing Gem, zip up the contents and place it in the "RemoteExampleV1" folder, and name it `RemoteExampleV1.zip`

Author: Ah, I'm not sure I understand what you mean - what is "it"?  The Gem?  So, something like, "When zipped the Gem is at the level where the gem.json is at the root of the zip file"?  I'm not sure I understand the last part - do you mean "When zipped, the Gem .zip archive is in the folder alongside the gem.json file"?

> 1. Create a Gem, or take an existing Gem, zip up the contents and place it in the "RemoteExampleV1" folder, and name it `RemoteExampleV1.zip`

Reviewer 2: It's just a small error I've seen a couple of times when giving similar instructions, where the folder was zipped so that that the root of the zip file was the gem folder itself and not the contents (for example gem.zip/ExampleGem/gem.json instead of gem.zip/gem.json). Though this is something we should also attempt to detect this in scripts.

> 1. Generate the sha256 for your Gem .zip archive and add that value to the "sha256" field in the `gem.json` file that is outside the .zip.  On Windows, you can use the `certutil` program to generate the SHA256 value like this:

Reviewer: ```suggestion
1. Generate the SHA-256 for your Gem's `.zip` archive and add that value to the `sha256` field of the `gem.json` file located outside the `.zip`.  On Windows, you can use the `certutil` program to generate the SHA-256 value like this:
```

    ```
    certutil -hashfile C:/path/to/gem.zip SHA256
    ```
    {{< note >}}
    It is possible to leave out the sha256 field for testing, but discouraged
    {{< /note >}}

> 1. Update your repo.json file and add the URL for the gem folder.  In this example we'll use the localhost URL for testing.

Reviewer: ```suggestion
1. Update your `repo.json` file and add the URL for the Gem folder.  This example uses the localhost URL for testing.
```

    ```
    {
        ...
        "gems": [
            "http://localhost:8080/RemoteExampleV1/"
        ]
    }
    ```

> ### Testing your local Gem Repository

Reviewer: ```suggestion
### Testing your local Gem repository
```

> To test your Gem Repository before you upload it to a server, you can host a local server using Python or register the folder path to your Gem Repository, but if you register the folder path, you'll need to modify the URIs to use absolute paths like `file://c/path/to/gem` instead of `http://localhost:8080/gem`

Reviewer: ```suggestion
To test your Gem repository before you upload it to a server, you can host a local server using Python or register the folder path to your Gem repository. If you register the folder path, you'll need to modify the URIs to use absolute paths like `file://c/path/to/gem` instead of `http://localhost:8080/gem`.
```

> 1. Use Python to host a local server for testing

Reviewer: ```suggestion
1. Use Python to host a local server for testing.
```

    ```
    cd path/to/gem/repository 
    
    # start a local python server on 8080
    python -m http.server 8080
    ```

> 1. Open up the Project Manager and add your Gem Repository, when prompted use the URL for your local server which should be "http://localhost:8080/"

Reviewer: ```suggestion
1. Open up Project Manager and add your Gem repository.  When prompted, use the URL for your local server, which should be `http://localhost:8080/`.
```

    {{< note >}}

>     Adding the Gem Repository may fail if you leave out the trailing slash in the URL.

Reviewer: ```suggestion
    Adding the Gem repository may fail if you leave out the trailing slash in the URL.
```

---
linkTitle: Overview
title: Gem Repository Overview
description: Gems repositories provide a way to share Gems in Open 3D Engine (O3DE).
weight: 100
toc: true
---

Gem repositories provide a way to discover, acquire, and share Gems that are not included with **Open 3D Engine (O3DE)**.  Gem repositories are generally hosted online, but you can also add them to your project from a local hard drive which is helpful for testing.

## Add a Gem repository in Project Manager

> You can add Gem repositories from the **Gem Repositories** page in Project Manager.

Reviewer 3: First, tell them where to find (how to navigate to) the Gem Repositories page.

> You can add Gem repositories from the **Gem Repositories** page in Project Manager which you can accessed from the **Engine** tab, or from the **Gem Catalog** right panel menu.

Reviewer: ```suggestion
You can add Gem repositories from the **Gem Repositories** page in Project Manager which you can access from the **Engine** tab, or from the **Gem Catalog** right panel menu.
```

1. Press the **Add Repository** button. In the dialogue box that opens you can supply the URL or local path for a Gem repository. 
1. If you have a URL for an online Gem repository, copy and paste it into the `Repository Path` field and press the **Add** button.

> 1. Alternately, if you want to add a Gem repository from your local hard drive, press the folder button to open the `Browse` dialog box, select the folder with the Gem repository, press the **Select Folder** button, and finally press the **Add** button.

Reviewer: ```suggestion
1. Alternatively, if you want to add a Gem repository from your local hard drive, press the folder button to open the `Browse` dialog box, select the folder with the Gem repository, press the **Select Folder** button, and finally press the **Add** button.
```

A new entry will be added to the list of Gem repositories, and the Gems provided by the Gem repository will be available to download and add to your projects.

## Add a Gem repository with the o3de command line tool

You can use the `o3de` CLI tool to add Gem repositories with the following command:
```
scripts\o3de.bat register -ru <Gem repository URL>
```

## Anatomy of a Gem repository  

A Gem repository contains a `repo.json` file which contains the metadata for the repository, including a list of URIs for the Gems it provides.

### Folder structure

The recommended folder structure is to put the `repo.json` file at the root of your Gem repository and the `gem.json` files for each Gem it provides in subfolders.  
```
/
   repo.json
   ExampleGem1/
       gem.json
   ExampleGem2/
       gem.json
```

If you are confident you will only ever have one Gem in the repository, it's acceptable to put the `gem.json` in the root folder.

> {{< note >}}
> One of the most convenient ways to host a Gem Repository online is to place the `repo.json` file inside the GitHub repository where your Gem's source code is.
> {{< /note >}}

Reviewer 3: This feels more like a "tip".
```suggestion
{{< tip >}}
One of the most convenient ways to host a Gem Repository online is to place the `repo.json` file inside the GitHub repository where your Gem's source code is.
{{< /tip >}}
```

### Format of a repo.json file

> The `repo.json` file contains the metadata for the Gem Repository and a list of Gems.

Reviewer 3: ```suggestion
The `repo.json` file contains the metadata for the Gem repository and a list of Gems.
```

> ```

Reviewer 3: ```suggestion
```json
```

{
    "repo_name":"RepositoryName",
    "origin":"RemoteOriginName",

>     "repo_uri": "https://o3de.org/repo/v1",

Reviewer 3: Nit: I realize you just want to use a fictional url for the repo, but I would think in the examples in these next two sections you'd want it to start with `https://github.com`, to help clarify the association with repos for the reader. So for example, `https://github.com/MyRepo/` for the repo_uri and `https://github.com/MyRepo/ExampleGem/` for the link to the Gem.

Author: Gem repos do not have to be in GitHub, and I wasn't sure at the time of writing if that was fully supported - though it looks like it might be?  If we were to use a GitHub URL example then we'd use `https://github.com/MyUsernameOrOrg/MyRepo` for the `repo_uri` and the Gem URL would probably be something like 
`https://github.com/MyUsernameOrOrg/MyRepo/releases/download/1.0.0/ExampleGem.zip`  (with SHA256 verification) or
`https://github.com/MyUsernameOrOrg/MyRepo/archive/refs/heads/main.zip` (without SHA256 verification)

    "summary": "Example Repository",
    "additional_info": "Example repository hosted on local http server.",
    "last_updated": "10/26/2021 04:56pm",
    "gems": [
        "http://localhost:8080/ExampleGem/"

>     ]

Reviewer 3: Since `localhost` is only used for testing, I don't think it should appear in the format examples. We should provide an example of the most common format - perhaps GitHub? So maybe use something like `"https://github.com/MyRepo"` and `"https://github.com/MyRepo/ExampleGem/"`.

}
```

In the future, the format of the `repo.json` file may be expanded to support projects, templates, and additional Gem repositories.

### Format of gem.json files 

The `gem.json` file for your Gems will need to be modified to include the `origin_uri`, `repo_uri`, `last_updated`, and `sha256` fields.

> ```

Reviewer 3: ```suggestion
```json
```

{
    "gem_name": "TestGem",
    "display_name": "Downloadable Test Gem",
    "license": "Apache-2.0 Or MIT",
    "origin": "Test Creator",
    "origin_uri": "http://localhost:8080/TestGem/gem.zip",

>     "repo_uri": "http://localhost:8080/", 

Reviewer 3: Same comment as above regarding usage of `localhost`.

}
```
| Field | Description |
| --- | --- |
| origin_uri | The URI for the `.zip` file containing the Gem archive.  This is a direct link to download the Gem's `.zip` archive. |
| repo_uri | The URI for the Gem repository. |
| last_updated | The date this file or Gem was last updated in `YYYY-MM-DD`, `YYYY-MM-DD HH:MM:SS`, or `YYYY-MM-DDTHH:MM:SS` format. |
| sha256 | The SHA-256 digest of the `.zip` archive that the `origin_uri` field points to.  You can omit this field for testing, but we highly recommend including it. |

{{< note >}}
For a full description of `gem.json` fields and examples see the [`gem.json` manifest documentation](/docs/user-guide/programming/gems/manifest/).
{{< /note >}}
## Creating a Gem repository using a template 

1. Create a new Gem repository using `o3de.bat`.  The following example uses a localhost URL as the repository URI for testing.

    ```
    scripts\o3de.bat create-from-template -tp Templates\GemRepo -dp <repository path on disk> -r ${RepoURI} http://localhost:8080/
    ```

1. Open the `repo.json` file in a text editor and update all the fields except the Gems list.  In the next step, you will update that when adding a Gem to the repository.

>     ```

Reviewer 3: ```suggestion
    ```json
```

    {
        "repo_name":"RepositoryName",
        "origin":"RemoteOriginName",
        "repo_uri": "http://localhost:8080/",
        "summary": "Example Repository",
        "additional_info": "Example repository hosted on local http server.",
        "last_updated": "10/26/2021 04:56pm",
        "gems": [
        ]
    }
    ```

### Add a Gem to the Gem repository

> 1.  Make a folder inside your Gem repository for the Gem.  In this example, the gem is named `RemoteExample`. Make a folder named `RemoteExampleV1` and create a `gem.json` file in the folder that looks like this:

Reviewer 3: ```suggestion
1.  Make a folder inside your Gem repository for the Gem.  In this example, the Gem is named `RemoteExample`. Make a folder named `RemoteExampleV1` and create a `gem.json` file in the folder that looks like this:
```

>     ```

Reviewer 3: ```suggestion
    ```json
```

    {
        "gem_name": "RemoteExample",
        "display_name": "Example Remote Gem",
        "license": "Apache-2.0 Or MIT",
        "origin":"RemoteOriginName",
        "origin_uri": "http://localhost:8080/RemoteExampleV1/RemoteExampleV1.zip",
        "sha256":"",
        "type": "Code",
        "summary": "An example remote gem.",
        "canonical_tags": [
            "Gem"
        ],
        "user_tags": [
            "Gameplay"
        ],
        "icon_path": "preview.png",
        "requirements": "",
        "documentation_url": "https://o3de.org/docs",
        "dependencies": []
    }
    ```

    The `origin_uri` points to the location of the Gem's `.zip` archive that will be downloaded.

1. Create a Gem, or take an existing Gem, zip up the Gem's contents, not the folder itself, and place the `.zip` in the "RemoteExampleV1" folder, and name it `RemoteExampleV1.zip`.

> {{< note >}}
> When creating an archive of your Gem, zip up the contents of the Gem folder, not the folder itself, so that when extracted, the `gem.json` file will be at the root, and not inside a subfolder named `RemoteExampleV1` 
> {{< /note >}}

Reviewer 3: ```suggestion
{{< tip >}}
When creating an archive of your Gem, zip up the contents of the Gem folder, not the folder itself, so that when extracted, the `gem.json` file will be at the root, and not inside a subfolder named `RemoteExampleV1` 
{{< /tip >}}
```

1. Generate the SHA-256 for your Gem's `.zip` archive and add that value to the `sha256` field of the `gem.json` file located outside the `.zip`.  On Windows, you can use the `certutil` program to generate the SHA-256 value like this:

> {{< note >}}

Reviewer 3: Move the note after the command line example. You'll need to indent the note shortcode when you do.

Author: I tried that but unfortunately it gets formatted incorrectly like code - possibly a Hugo bug? To work around it I avoid indenting any notes or tips.

Reviewer: The opening tag is indented the expected amount (here, four spaces, to match the indentation of the code block).  The indentation level of the tag's body is what causes issues.  _Any_ indentation in a tag's body is treated as a new level.  Indenting the body to match the opening tag will create something that looks like a codeblock inside of the note.  Instead, remove all indentation from the body (the closing tag's indentation level doesn't matter).  I'll suggest this exact change for clarity.

It is possible, but not recommended, to leave out the SHA-256 field for testing.
{{< /note >}}
    ```
    certutil -hashfile C:/path/to/gem.zip SHA256
    ```
1. Update your `repo.json` file and add the URL for the Gem folder.  This example uses the localhost URL for testing.

>     ```

Reviewer 3: ```suggestion
    ```json
```

> {{< note >}}
> It is possible, but not recommended, to leave out the SHA-256 field for testing.
> {{< /note >}}
>     ```
>     certutil -hashfile C:/path/to/gem.zip SHA256
>     ```

Reviewer: ```suggestion
    ```
    certutil -hashfile C:/path/to/gem.zip SHA256
    ```
    {{< note >}}
It is possible, but not recommended, to leave out the SHA-256 field for testing.
{{< /note >}}
```
Copied from a resolved conversation:

The opening tag is indented the expected amount (here, four spaces, to match the indentation of the code block). The indentation level of the tag's body is what causes issues. Any indentation in a tag's body is treated as a new level. Indenting the body to match the opening tag will create something that looks like a codeblock inside of the note. Instead, remove all indentation from the body (the closing tag's indentation level doesn't matter). I'll suggest this exact change for clarity.

Author: Excellent, not indenting the rest of the tag contents worked.
